client/components/domains/registrant-extra-info/with-contact-details-validation.jsx

```diff
@@ -73,20 +76,68 @@ export function formatIMJVErrors( errors, schema ) {
 	);
 }
 
-const WithContactDetailsValidation = ( schema, WrappedComponent ) => {
-	return class FormWithValidation extends Component {
-		validate = validatorFactory( schema, { greedy: true, verbose: true } );
+export default function WithContactDetailsValidation( tld, WrappedComponent ) {
+	const wrappedComponentName = WrappedComponent.displayName || WrappedComponent.name || '';
+
+	class ComponentWithValidation extends Component {
+		static propTypes = {
+			validationSchema: PropTypes.object.isRequired,
+		};
+
+		displayName = 'WithContactDetailsValidation(' + tld + ', ' + wrappedComponentName + ')';
+
+		validate = () => {};
+
+		componentWillMount() {
+			this.compileValidator();
+		}
+
+		componentWillReceiveProps( nextProps ) {
+			if ( nextProps.validationSchema !== this.props.validationSchema ) {
+				this.compileValidator();
+			}
+		}
+
+		compileValidator() {
+			// The schema object we compile is available at this.validate.toJSON()
+			// but we already check it's different in componentWillReceiveProps
+			debug( `compiling validation schema for ${ tld }`, this.props.validationSchema );
+			this.validate = validatorFactory( this.props.validationSchema, {
+				greedy: true,
+				verbose: true,
+			} );
+		}
 
 		validateContactDetails() {
-			const isValid = this.validate( this.props.contactDetails );
+			if ( typeof this.validate !== 'function' ) {
+				return {};
+			}
+
+			// This is pretty fast, but is a candidate for memoization
+			debug( 'validating contactDetails' );
+			let isValid = false;
+			try {
+				isValid = this.validate( this.props.contactDetails );
+			} catch ( error ) {
+				// An exception here usually indicates a problem with the schema
+				// but could result from malformed contactDetails.
+				// We console.error here so it gets logged and we can track it.
+				// eslint-disable-next-line no-console
+				console.error( 'Error thrown during validation:', {
+					error,
+					tld: tld,
+					contactDetails: this.props.contactDetails,
+				} );
+				return { '': [ 'There was an unexpected error validating these details' ] };
```

Comment:
this empty key demands an explanation - why is it there? (please add a comment in the code :wink:)